### ISSUE TITLE
Self describing Pipeline and Task Definitions

### ISSUE THREAD

**Issue by objectiveous** [tektoncd/pipeline#993]
# Overview 

A number of teams are building catalogs of Pipelines and Tasks for the purpose of reuse. Many of these catalog have usage documentation that exists separate from the the actual Pipeline/Task definition. 

# Proposal 

Make usage documentation a fist class concern of Tasks. Perhaps by enhancing task definitions or maybe by introducing a new Kind (TaskInfo?). 

# Example Scenario 

Consider a team that uses a tekton catalog to populate the Design Pallet of CI/CD GUI tool. Having populated a Design Pallet, users may want to Inspect a task to lean more about it's Inputs, Outputs, version number, intended use, etc.

**Comment by bobcatfish** [comment]
This would be great to see! In the catalog repo we're ending up with a lot of docs in the associated readmes (e.g. https://github.com/tektoncd/catalog/tree/master/golang#golangci-lint) but I agree it would be way better to see this in the Tasks and Pipelines themselves!

Also maybe related: https://github.com/tektoncd/catalog/issues/7

**Comment by bobcatfish** [comment]
I want to add here that `PipelineResources` should have descriptions too!

**Comment by vdemeester** [comment]
Moved it to beta as I feel it can be a small change (doesn't even need to be an API change), and it would be a really really nice to have for beta.

**Comment by piyush-garg** [comment]
@bobcatfish @vdemeester 

What should we add regarding this in pipelines? Should we add an optional `Description` field in TaskSpec? Or any other approach?

**Comment by siamaksade** [comment]
+1 on adding `Description` to TaskSpec

**Comment by bobcatfish** [comment]
hey there! i think this is being discussed in slack also but I noticed that #2089 is doing this via annotation, i dont quit understand why we chose annotation vs. a field in the spec?

i would prefer an annotation if possible, a couple reasons:
- other fields like this would be done via spec (e.g. description of a param)
- adding it to the spec makes it part of the API - imo using the annotation is a kubernetes specific thing, i think we still want to at least entertain the possibility of supporting a non k8s use case
